fs/wildcard.py

```diff
@@ -123,55 +135,92 @@ def get_matcher(patterns, case_sensitive):
     if not patterns:
         return lambda name: True
     if case_sensitive:
-        return partial(match_any, patterns)
+        return lambda name: match_any(patterns, name, accept_prefix)
     else:
-        return partial(imatch_any, patterns)
+        return lambda name: imatch_any(patterns, name, accept_prefix)
+
+
+class _PatternMatcher:
+    def __init__(self, pattern, case_sensitive, accept_prefixes):
+        self.case_sensitive = case_sensitive
+        self.accept_prefixes = accept_prefixes
+
+        self.tokens = []
+        i = 0
+        while i < len(pattern):
+            if pattern[i] == "[":
+                start = i
+                i = pattern.find("]", i) + 1
+                self.tokens.append(_PatternMatcherToken(pattern[start:i]))
+            elif pattern[i] == "]":
+                raise ValueError(
+                    pattern + " is not a valid wildcard pattern. (unmatched ])"
+                )
+            elif pattern[i] == "?":
+                self.tokens.append(_PatternMatcherToken("?"))
+                i += 1
+            elif pattern[i] == "*":
+                asterik_len = 1
+                while (asterik_len + i) < len(pattern) and pattern[
+                    asterik_len + i
+                ] == "*":
+                    asterik_len += 1
+                if asterik_len > 2:
+                    raise ValueError(
+                        pattern
+                        + " is not a valid wildcard pattern. "
+                        + "(sequence of three or more *)"
+                    )
```

Comment:
Seeing these different errors, I'm tempted to add a `PatternError` to `fs.errors` here, and maybe subclass it (either `TooManyTokensError` or `UnmatchedTokenError`).

Reply:
I added a `PatternError` which also stores the position at which the error occurred. That seemed nicer to me than having a new class for each type of error but tell me if that isn't what you had in mind.